src/data/image.ts

```diff
@@ -35,29 +66,70 @@ const streamToS3 = (
   return pass;
 };
 
-export async function resizeImage(
-  imageUrl: URL,
-  size: number
+export async function uploadUserImage(
+  userSlug: string,
+  imageUrl: URL
 ): Promise<string> {
-  const s3 = new S3({
-    accessKeyId: process.env.ACCESS_KEY_ID,
-    secretAccessKey: process.env.SECRET_ACCESS_KEY,
-    region: process.env.REGION,
-  });
+  const bucketPrefix = "team-full/";
+  const Bucket = process.env.IMAGE_BUCKET as string;
 
   if (!imageUrl.href.startsWith("https://drive.google.com")) {
     throw new Error(`Error processing ${imageUrl.href}, not a drive url`);
   }
 
   const imageId = imageUrl.href.split("/file/d/")[1]?.split("/")[0];
-
   if (!imageId) {
     throw new Error(
       `Invalid formatted google drive image url: ${imageUrl.href}`
     );
   }
 
-  const resizer = sharp().rotate().resize(size, size);
+  /**
+   * Used to determine the uniqueness of images. Any images w/ the same
+   * hash for the same user won't be overwrote. If changes to the image
+   * processing are done you can update the salt value to regenerate all
+   * the hashes.
+   */
+  const imageUrlHash = hash(imageUrl.toString() + "salt2");
+  const userImageName = `${bucketPrefix}${userSlug}-${imageUrlHash}`;
+
+  const [ErrorAccessingS3Images, S3Images] = await to(
+    listS3Images(bucketPrefix)
+  );
+  if (ErrorAccessingS3Images || !S3Images) {
+    log.error("Unable to access images on S3", ErrorAccessingS3Images);
+  } else {
+    const userImages: string[] =
+      S3Images.Contents?.map((image) => image.Key ?? "").filter((image) =>
+        image.includes(`/${userSlug}-`)
+      ) ?? [];
+    const [currentUserImage, oldUserImages] = extractFirstPartialMatch(
+      userImageName,
+      userImages
+    );
+    // Delete any old images
+    for (let oldImage of oldUserImages) {
+      const [deleteError] = await to(
+        s3
+          .deleteObject({
+            Bucket,
+            Key: oldImage,
+          })
+          .promise()
+      );
+      if (deleteError) {
+        log.error(`Failed to delete old user image ${oldImage}`, deleteError);
+      }
+    }
+    if (currentUserImage) {
+      // This image has already been uploaded, let's skip the rest
+      return `https://${Bucket}.s3.amazonaws.com/${currentUserImage}`;
+    }
+  }
+
+  log.info(`Uploading image for ${userSlug}`);
+
+  const normalizer = sharp().rotate().resize(500, 500);
```

Comment:
There's a lot of magic in this one line. It normalizes the image sizes and corrects their orientation. Without it some folks would be sideways and a lot of folks would be squashed. Giving both sizes is super important as providing only one still ends up with squashing. Sharp does a decent job of finding a face and cropping appropriately.